### ISSUE TITLE
[RFC] NKI Kernels

### ISSUE THREAD

**Issue by rthekini-aws** [pytorch/torchtitan#2391]
# RFC: NKI Kernel Alternatives for AWS Trainium Support

## Summary

Provide [NKI](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/nki/index.html) (Neuron Kernel Interface) alternatives to the existing Triton kernels in `torchtitan`, enabling training on AWS Trainium instances. Existing Triton and CPU paths remain untouched — NKI kernels are selected only when running on Neuron devices.

## Motivation

`torchtitan` uses Triton kernels for performance-critical operations (e.g., `_fill_indices_kernel` in `torchtitan/models/moe/kernels.py` for MoE token permutation). Triton is primarily catered towards GPUs (with a focus on NVIDIA) and is not available on AWS Trainium hardware.

NKI is a Python-based DSL that provides direct access to the Neuron Instruction Set Architecture (NISA), analogous to how Triton provides low-level GPU kernel authoring. NKI kernels can be inserted into PyTorch models as custom operators, making them a natural drop-in alternative for Triton kernels when running on Neuron devices.

By providing NKI equivalents of the existing Triton kernels, `torchtitan` can run on Trainium without modifying any GPU code paths.

### In scope
- NKI implementations of existing Triton kernels in `torchtitan` (currently `_fill_indices_kernel` for MoE permute indices)
- A device-aware dispatch mechanism that selects the appropriate kernel at runtime
- Future NKI kernels for additional operations as `torchtitan` adds more Triton kernels

### Out of scope
- Changes to model architecture, training loop, or parallelization logic
- Replacing or modifying existing Triton kernels
- Neuron-specific distributed strategies or device management

### Current state

```
torchtitan/models/moe/kernels.py
├── _fill_indices_kernel()      # Triton kernel
├── fill_indices_wrapper()      # Triton launcher
├── fill_indices_cpu()          # CPU reference implementation
└── generate_permute_indices()  # Public API (dispatches to Triton or CPU)
```

`generate_permute_indices()` already has a `use_cpu` flag for dispatch. The NKI approach extends this pattern.

### Proposed structure

```
torchtitan/models/moe/
├── kernels.py                  # Existing file, modified with dispatch logic
└── nki_kernels.py              # New file: NKI kernel implementations
```

### Dispatch mechanism

The dispatch in `generate_permute_indices()` gains a Neuron path, selected automatically based on the input tensor's device:

```python
# In kernels.py — generate_permute_indices()

if use_cpu:
    permuted_indices = fill_indices_cpu(...)
elif tokens_per_expert_group.device.type == "neuron":
    from .nki_kernels import fill_indices_nki
    permuted_indices = fill_indices_nki(...)
else:
    permuted_indices = fill_indices_wrapper(...)  # existing Triton path
```

The lazy import ensures Neuron SDK packages are only required when actually running on Neuron hardware. No new dependencies are added for GPU users.

### NKI kernel implementation

NKI kernels are authored using the [NKI language](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/nki/get-started/nki-language-guide.html) and registered as PyTorch custom operators via `torch-neuronx`.

At a high level, the NKI kernel file (`nki_kernels.py`) will:

1. Define the kernel logic using `nki.jit` and NKI language primitives (tile-based loads, stores, and compute)
2. Register the kernel as a PyTorch custom op via `torch-neuronx`
3. Expose a `fill_indices_nki()` wrapper with the same signature as `fill_indices_wrapper()`, so the dispatch in `kernels.py` is a simple swap

### torch.compile compatibility

NKI kernels registered as custom ops are opaque to `torch.compile`. They work with the `neuron` Dynamo backend (`torch.compile(model, backend="neuron")`), which is the expected compilation path on Trainium. No special handling is needed in `torchtitan`'s compile configuration.

## Relationship to `torchtitan` experiments guidelines

Note: this change targets `torchtitan/models/moe/`, not the `experiments/` folder. However, the experiments guidelines are relevant because they articulate `torchtitan`'s stance on vendor-specific code:

> An experiment should not contain vendor-specific code, such as kernels written in a proprietary language. Those can be hosted outside as dependency.

This RFC proposes a minimal approach: the NKI kernel implementations live in a single new file (`nki_kernels.py`) that is lazy-imported only on Neuron devices. The actual vendor-specific dependencies (`neuronxcc`, `torch-neuronx`) are external packages installed on Neuron instances — they are not vendored into `torchtitan`. The dispatch logic added to `kernels.py` is a few lines of device-type checking, consistent with how the existing `use_cpu` flag works.

If the `torchtitan` maintainers prefer stricter separation, the NKI kernels could alternatively be shipped as an external pip package. However, this adds friction and makes it harder to keep kernel signatures in sync with `torchtitan` changes.

## Dependencies

When running on Neuron hardware (already present on Neuron instances):
- `neuronxcc` — Neuron compiler, provides the NKI language (`nki`, `nki.language`, `nki.isa`)
- `torch-neuronx` — PyTorch backend for Neuron, provides custom op registration for NKI kernels

No changes to `requirements.txt`. These are not required for GPU or CPU users.

## Testing

- NKI kernel outputs validated against the existing `fill_indices_cpu` reference implementation (same approach used by the existing Triton kernel tests in `tests/unit_tests/test_permute_indices_kernel.py`)
- Tests gated on Neuron device availability (`@unittest.skipUnless`)
- GPU and CPU paths remain completely unaffected

## Rollout

1. Add `nki_kernels.py` with the `fill_indices` NKI kernel
2. Add device-based dispatch in `generate_permute_indices()`
3. Add gated unit tests
4. As `torchtitan` adds more Triton kernels, provide NKI equivalents following the same pattern

**Comment by tianyu-l** [comment]
I think we can deprecate `fill_indices_cpu` after https://github.com/pytorch/torchtitan/pull/2255 land. In particular, the `generate_permute_indices` can be pure python / pytorch, since we no longer need to support padding.

cc @rakkit @shuhuayu @danielvegamyhre

**Comment by rthekini-aws** [comment]
If I'm understanding that PR correctly, it just changes the BF16 path, so a triton kernel may still need to be used on the other paths, right?

Additionally, I seem to be missing some context here, since I believe that the BF16 `_grouped_mm` CUDA implementation still needs padding for TMA: https://github.com/pytorch/pytorch/blob/1a5e4f63338006744610985167112dba0cf20406/aten/src/ATen/native/cuda/GroupMMCommon.cuh#L64-L67

**Comment by tianyu-l** [comment]
For mxfp8 path, @danielvegamyhre and I are discussing how to move the padding to torchao, instead of in torchtitan.

For TMA, cc @ngimel

**Comment by rakkit** [comment]
yes. without padding, `generate_permute_indices` can be implemented with pure pytorch. Draft implement [here](https://github.com/rakkit/torchtitan/blob/moe_fast_path/torchtitan/models/moe/utils.py#L11)
